core/test/base/batch_lin_op.cpp

```diff
@@ -0,0 +1,413 @@
+        ref, std::vector<gko::dim<2>>{gko::dim<2>{3, 4}});
+
+    ASSERT_THROW(op2->apply(gko::lend(wrong), gko::lend(x2)),
+                 gko::DimensionMismatch);
+}
+
+
+TEST_F(EnableBatchLinOp, ApplyFailsOnWrongSolutionRows)
+{
+    auto wrong = DummyBatchLinOp::create(
+        ref, std::vector<gko::dim<2>>{gko::dim<2>{5, 4}});
+
+    ASSERT_THROW(op->apply(gko::lend(b), gko::lend(wrong)),
+                 gko::DimensionMismatch);
+}
+
+
+TEST_F(EnableBatchLinOp, ApplyFailsOnOneBatchWrongSolutionRows)
+{
+    auto wrong = DummyBatchLinOp::create(
+        ref, std::vector<gko::dim<2>>{gko::dim<2>{5, 4}, gko::dim<2>{5, 4}});
+
+    ASSERT_THROW(op2->apply(gko::lend(b2), gko::lend(wrong)),
+                 gko::DimensionMismatch);
+}
+
+
+TEST_F(EnableBatchLinOp, ApplyFailsOnWrongSolutionColumns)
+{
+    auto wrong = DummyBatchLinOp::create(
+        ref, std::vector<gko::dim<2>>{gko::dim<2>{3, 5}});
+
+    ASSERT_THROW(op->apply(gko::lend(b), gko::lend(wrong)),
+                 gko::DimensionMismatch);
+}
+
+
+TEST_F(EnableBatchLinOp, ExtendedApplyFailsOnWrongBSize)
+{
+    auto wrong = DummyBatchLinOp::create(
+        ref, std::vector<gko::dim<2>>{gko::dim<2>{3, 4}});
+
+    ASSERT_THROW(op->apply(gko::lend(alpha), gko::lend(wrong), gko::lend(beta),
+                           gko::lend(x)),
+                 gko::DimensionMismatch);
+}
+
+
+TEST_F(EnableBatchLinOp, ExtendedApplyFailsOnWrongSolutionRows)
+{
+    auto wrong = DummyBatchLinOp::create(
+        ref, std::vector<gko::dim<2>>{gko::dim<2>{5, 4}});
+
+    ASSERT_THROW(op->apply(gko::lend(alpha), gko::lend(b), gko::lend(beta),
+                           gko::lend(wrong)),
+                 gko::DimensionMismatch);
+}
+
+
+TEST_F(EnableBatchLinOp, ExtendedApplyFailsOnWrongSolutionColumns)
+{
+    auto wrong = DummyBatchLinOp::create(
+        ref, std::vector<gko::dim<2>>{gko::dim<2>{3, 5}});
+
+    ASSERT_THROW(op->apply(gko::lend(alpha), gko::lend(b), gko::lend(beta),
+                           gko::lend(wrong)),
+                 gko::DimensionMismatch);
+}
+
+
+TEST_F(EnableBatchLinOp, ExtendedApplyFailsOnWrongAlphaDimension)
+{
+    auto wrong = DummyBatchLinOp::create(
+        ref, std::vector<gko::dim<2>>{gko::dim<2>{2, 5}});
+
+    ASSERT_THROW(op->apply(gko::lend(wrong), gko::lend(b), gko::lend(beta),
+                           gko::lend(x)),
+                 gko::DimensionMismatch);
+}
+
+
+TEST_F(EnableBatchLinOp, ExtendedApplyFailsOnWrongBetaDimension)
+{
+    auto wrong = DummyBatchLinOp::create(
+        ref, std::vector<gko::dim<2>>{gko::dim<2>{2, 5}});
+
+    ASSERT_THROW(op->apply(gko::lend(alpha), gko::lend(b), gko::lend(wrong),
+                           gko::lend(x)),
+                 gko::DimensionMismatch);
+}
+
+
+// For tests between different memory, check cuda/test/base/batch_lin_op.cu
+TEST_F(EnableBatchLinOp, ApplyDoesNotCopyBetweenSameMemory)
+{
+    op->apply(gko::lend(b), gko::lend(x));
+
+    ASSERT_EQ(op->last_b_access, ref);
```

Comment:
This is not clear to me. How do we know that no copy happened from this assertion?

Reply:
Because `op` is created on `ref2` and on on `ref`. Therefore we check that b is not copied to `ref2` because `ref` and `ref2` have the same memory space.